_posts/2018-11-29-async-update.md

```diff
@@ -0,0 +1,201 @@
+```
+
+This synchronous method *blocks the current thread* until data has been read
+into `buf`, then says how many bytes were read. We can build on this method
+to implement `read_exact`, a method that *continues* reading until the buffer
+is filled:
+
+```rust
+fn read_exact<T: Read>(input: &mut T, buf: &mut [u8]) -> Result<(), std::io::Error> {
+    let mut cursor = 0;
+    while cursor < buf.len() {
+        cursor += input.read(&mut buf[cursor..])?;
+    }
+}
+```
+
+In the asynchronous world, we want to perform similar operations, but rather
+than blocking the current thread we want to leave it free to do other work
+while the I/O operations complete asychronously. But actually programming
+*directly* in that way is incredibly difficult. What we want is to program
+*as if* I/O operations will block the current thread, but have supporting
+infrastructure transform this code into more efficient asynchronous execution.
+
+In short, our goal is to be able to write the following:
+
+```rust
+async fn read_exact<T: AsyncRead>(input: &mut T, buf: &mut [u8]) -> Result<(), std::io::Error> {
+    let mut cursor = 0;
+    while cursor < buf.len() {
+        cursor += await!(input.read(&mut buf[cursor..]))?;
+    }
+}
+```
+
+Comparing the two snippets, there are three changes here:
+
+- We write `async` before `fn`, to signal that the function should be
+asynchronously executed on its parent thread. Async functions return
+their result within a `Future`, representing a value that must be asynchronously
+computed.
+
+- We use the `AsyncRead` trait (from `futures::io`), rather than the
+`Read` trait. This makes an asynchronous version of the `read` method
+available (curretly provided via the `AsyncReadExt` extension trait).
+
+- We enclose the call to `read` with `await!`, signaling that we want to
+*simulate* blocking on the operation to complete.
+
+And that's all.
+
+This kind of systematic transformation is one that has proved itself
+repeatedly in other languages. But there's an extra element in the Rust
+version: borrowing. For `read_exact`, we are able to hold the borrows
+of `input` and `buf` while we use `await!` (which may actually clear
+the stack and run completely unrelated code). The validity of the
+borrowing is still checked, and it works largely similarly to borrowing
+within synchronous code.
+
+If you've programmed with futures in Rust before, you'll know this is a
+game changer: manual futures code generally must be restricted to `'static`
+data, which (in addition to its verbosity) takes it far away from
+idiomatic Rust.
+
+# The book
+
+Part of the work around async/await this year has been writing a new
+book, covering the syntax, the underlying `Future` API, and ultimately
+various programming patterns that emerge. @cramertj has written [an early
+draft][apr], which is already useful for understanding these concepts.
+
+[apr]: https://github.com/rust-lang-nursery/wg-net/blob/master/async-book/src/TOC.md
```

Comment:
IMO we should build and publish a rendered version of this rather than linking to the GH markdown. Want to make a GH pages repo, and we can set up a hook to publish to it?